gem/optimise.py

```diff
@@ -413,3 +480,565 @@ def aggressive_unroll(expression):
+    mcf_value = [0, 1]  # (num arg indices, count)
+    mcf = None
+    arg_ind_set = set(arg_ind_flat)
+    for factor, count in counter.items():
+        if count > 1:
+            num_arg_ind = len(set(factor.free_indices) & arg_ind_set)
+            if (num_arg_ind > mcf_value[0]) or (num_arg_ind == mcf_value[0] and count > mcf_value[1]):
+                # prioritize factors with more argument indices
+                mcf = factor
+                mcf_value[0] = num_arg_ind
+                mcf_value[1] = count
+
+    if not mcf:
+        # no common factors
+        return factor_lists
+
+    new_list = []
+    rest = []  # new_list = mcf * [rest] + rest
+    # keep original factor list
+    # this might not be economical as most of the times factorise should help,
+    # and the copying slows down the optimisation quite a bit
+    # before_rest = []
+    for fl in factor_lists:
+        if mcf in fl and len(fl) > 1:
+            # at least two terms in the product
+            # before_rest.append(fl)
+            # new_fl = list(fl)
+            # new_fl.remove(mcf)
+            # rest.append(new_fl)
+            fl.remove(mcf)
+            rest.append(fl)
+        else:
+            new_list.append(fl)
+
+    # before_node = reorder(sumproduct_2_node(before_rest), (quad_ind + arg_ind_flat))
+    # before_flop = count_flop(before_node)
+    # after_node = reorder(sumproduct_2_node(factorise(rest, quad_ind, arg_ind_flat)), (quad_ind + arg_ind_flat))
+    # after_flop = count_flop(Product(mcf, after_node))
+    # if after_flop > before_flop:
+    #     # only factorise if reduces flops
+    #     return factor_lists
+    # new_list.append([mcf, after_node])
+    new_list.append([mcf, sumproduct_2_node(factorise(rest, quad_ind, arg_ind_flat))])
+    return factorise(new_list, quad_ind, arg_ind_flat)  # recursion
+
+
+def get_array(tensor, subarray=None):
+    if isinstance(tensor.children[0], Identity):
+        dim = tensor.children[0].dim
+        if subarray:
+            return eval('numpy.identity({0})'.format(dim) + subarray)
+        else:
+            return numpy.identity(dim)
+    if subarray:
+        return eval('tensor.children[0].array' + subarray)
+    else:
+        return tensor.children[0].array
+
+
+def contract(tensors, free_indices, indices):
+    """
+    :param tensors: (A, B, C, ...)
+    :param free_indices: contract over (i, ...)
+    :param indices: (j, k, ...)
+    :return:
+    """
+    index_map = {}
+    letter = ord('a')
+    for i in indices + free_indices:
+        index_map[i] = chr(letter)
+        letter += 1
+    subscripts = []
+    arrays = []
+    for t in tensors:
+        if any(isinstance(i, int) for i in t.multiindex):
+            subarray = [str(i) if isinstance(i, int) else ':' for i in t.multiindex]
+            subarray = '[' + ','.join(subarray) + ']'  # e.g. [:,:,0]
+            # this bit is a bit ugly
+            arrays.append(get_array(t, subarray))
```

Comment:
Oh christ, you're using this to do indexing of arrays by concatenating strings and then calling eval?!

You want:

```
idx = [i if isinstance(i, int) else slice(None) for i in t.multiindex]
arrays.append(t.array[idx])
```